### ISSUE TITLE
🖊️ Color in print

### ISSUE THREAD

**Issue by Felienne** [hedyorg/hedy#5836]
Idea from a student, can we also have colors in the output? I guess disconnected from Python/Skulpt we can do that with realitive ease?

**Comment by jpelay** [comment]
Yes, we totally can, the function that prints to the output is controlled by us, so its just a matter of setting the color. The complicated part would to ideate a way to communicate with this function from Python, but we already call a bit of JavaScript from Python

https://github.com/hedyorg/hedy/blob/e8460efdfddf11f5927595d15d2b222433cdc3fc/static/js/app.ts#L1166-L1176

**Comment by jpelay** [comment]
Do you have a command in mind or maybe make the color customizable within the UI?

**Comment by Felienne** [comment]
I think the kids expectation was that:

```
color blue
print Hello!
```

Would print Hello! in blue which I think is a reasonable expectation?

So it was not so much an idea of a kids as a misconception, haha!

**Comment by jpelay** [comment]
> ```
> color blue
> print Hello!
> ```

I think that makes sense! I think a bit of control on the output could be a lot of fun!

**Comment by Kache** [comment]
I suggest leading students toward the concept of markup, i.e. storing color information along with the value being printed, would be a much better idea than having a "secretly stateful colored print". Instead of going into my reasons, straight to example:

```
color blue
name is Sophie
color red
fruit is apples
print name likes fruit
```

And to demonstrate, a behaviorally valid transpilation for the above that renders HTML:

```python
def markup(val: str, color: str | None):
    if color:
        return '<span style="color:{color}">{val}</span>'
    return val

curr_color = 'blue'
name = markup('Sophie', color=curr_color)
curr_color = 'red'
fruit = markup('apples', color=curr_color)
print(f"{name} likes {fruit}")
```

**Comment by Kache** [comment]
Another idea, trying to get away from color commands that depend on some global "color state":

```
# prereq: quoting
name is blue 'Sophie'
fruit is red 'apples'
print name likes fruit
```

Which we later reveal as plain static function calls:

```
# prereq: parethesis function calls
name = blue('Sophie')
fruit = red('apples')
print f'{name} likes {fruit}'
```

**Comment by Kache** [comment]
Got a start on this. Idea: implement `color` commands as printing [ANSI color escape codes](https://en.wikipedia.org/wiki/ANSI_escape_code#Colors), then later converting them to styled HTML tags on the FE

This following patch does the first part: https://github.com/hedyorg/hedy/compare/26.3.13...Kache:color-print
Commits are organized atomically for ease of viewing commit-by-commit

Thoughts on this approach? Would appreciate some early input @jpelay @AnneliesVlaar

**Comment by Kache** [comment]
@jpelay great! and also thanks!

I've taken a look at the frontend implementation as well, now. I'll work on getting it running locally and put up a PR.

**Comment by Kache** [comment]
moving discussion to https://github.com/hedyorg/hedylang/issues/2 for now, since compiler part of the code is not yet complete

Can continue here with FE implementation once it is.

**Details**

The [previously posted draft](https://github.com/hedyorg/hedy/issues/5836#issuecomment-4078817153) implements text color by transpiling `color` commands into standalone Python lines that `print()` ASCII color escape codes. Although that works in "pure hedylang" isolation, it translates poorly to the JS/DOM-based web environment, see:

https://github.com/hedyorg/hedy/blob/27bd4cd92de0f5e4d3d55c39e225bd7c6db4890c/static/js/app.ts#L1140-L1150

It'd be necessary to implement an analogue of the terminal's streaming-escape-code-printing, translating it into the HTML DOM world by detecting escape codes and tracking "current color state" in JS as text is added to the DOM.

So, I'm revisiting a hedylang implementation that prefixes and postfixes the escape codes in the `print()` call as a complete "start color" and "end color" unit of marked-up text, which will translate much more easily when it gets to the FE

**Comment by jpelay** [comment]
Hi, Kevin, sorry I missed this message before.

> So, I'm revisiting a hedylang implementation that prefixes and postfixes the escape codes in the `print()` call as a complete "start color" and "end color" unit of marked-up text, which will translate much more easily when it gets to the FE

This makes a lot of sense! A powerful idea is that we can mix JavaScript when transpiling as well. This means tangling the compiler to the front end more, but is something that we have done in the past as well.
